src/main/java/org/openjdk/jextract/json/JSONObjects.java

```diff
@@ -0,0 +1,826 @@
+        if (targetType == boolean.class ||
+            targetType == Boolean.class ||
+            targetType == Object.class) {
+            return jb.asBoolean();
+        } else {
+            throw cannotConvert(jb, targetType);
+        }
+    }
+
+    private static final Map<Class<?>, Function<JSONNumber, Object>> numberConverters =
+        new IdentityHashMap<>();
+
+    private static void addNumberConverter(Class<?> cls, Function<JSONNumber, Object> func) {
+        numberConverters.put(cls, func);
+    }
+
+    static {
+        addNumberConverter(byte.class, JSONNumber::asByte);
+        addNumberConverter(Byte.class, JSONNumber::asByte);
+        addNumberConverter(short.class, JSONNumber::asShort);
+        addNumberConverter(Short.class, JSONNumber::asShort);
+        addNumberConverter(int.class, JSONNumber::asInt);
+        addNumberConverter(Integer.class, JSONNumber::asInt);
+        addNumberConverter(long.class, JSONNumber::asLong);
+        addNumberConverter(Long.class, JSONNumber::asLong);
+        addNumberConverter(OptionalInt.class, jn -> OptionalInt.of(jn.asInt()));
+        addNumberConverter(OptionalLong.class, jn -> OptionalLong.of(jn.asLong()));
+        addNumberConverter(BigInteger.class, jn -> BigInteger.valueOf(jn.asLong()));
+        addNumberConverter(BigDecimal.class, jn -> BigDecimal.valueOf(jn.asLong()));
+        Function<JSONNumber, Object> intOrLong = jn -> {
+            try {
+                 return jn.asInt();
+            } catch (ArithmeticException ae) {
+                 return jn.asLong();
+            }
+        };
+        addNumberConverter(Number.class, intOrLong);
+        addNumberConverter(Object.class, intOrLong);
+    }
+
+    // convert the JSONNumber value to a Java object of the given Type
+    private static Object convertNumber(JSONNumber jn, Type targetType) {
+        var func = numberConverters.get(targetType);
+        if (func != null) {
+            return func.apply(jn);
+        } else {
+            throw cannotConvert(jn, targetType);
+        }
+    }
+
+    private static final Map<Class<?>, Function<JSONDecimal, Object>> decimalConverters =
+        new IdentityHashMap<>();
+
+    private static void addDecimalConverter(Class<?> cls, Function<JSONDecimal, Object> func) {
+        decimalConverters.put(cls, func);
+    }
+
+    static {
+        addDecimalConverter(double.class, JSONDecimal::asDouble);
+        addDecimalConverter(Double.class, JSONDecimal::asDouble);
+        addDecimalConverter(Number.class, JSONDecimal::asDouble);
+        addDecimalConverter(Object.class, JSONDecimal::asDouble);
+        addDecimalConverter(float.class, JSONDecimal::asFloat);
+        addDecimalConverter(Float.class, JSONDecimal::asFloat);
+        addDecimalConverter(OptionalDouble.class, jd -> OptionalDouble.of(jd.asDouble()));
+        addDecimalConverter(BigDecimal.class, jd -> BigDecimal.valueOf(jd.asDouble()));
```

Comment:
Not sure we'll ever use this one TBH?

Reply:
Removed BigDecimal. float exists only for completeness of 'all primitive' types.